libraries/browser-client/src/websocket/worker.ts

```diff
@@ -0,0 +1,383 @@
+import AudioProcessor from '../audioprocessing/AudioProcessor'
+import EnergyTresholdVAD from '../audioprocessing/EnergyTresholdVAD'
+import AudioTools from '../audioprocessing/AudioTools'
+import { ControllerSignal, WebsocketResponseType, WorkerSignal } from './types'
+import { AudioProcessorParameters, ContextOptions, VadOptions } from '../client'
+
+/**
+ * The interface for response returned by WebSocket client.
+ * @public
+ */
+interface WebsocketResponse {
+  /**
+   * Response type.
+   */
+  type: WebsocketResponseType
+
+  /**
+   * Audio context ID.
+   */
+  audio_context: string
+
+  /**
+   * Segment ID.
+   */
+  segment_id: number
+
+  /**
+   * Response payload.
+   *
+   * The payload value should match the response type (i.e. TranscriptResponse should have Transcript type).
+   * Not all response types have payloads - Started, Stopped and SegmentEnd don't have payloads.
+   * TentativeIntent and Intent share the same payload interface (IntentResponse).
+   */
+  data: any
+}
+
+const CONTROL = {
+  WRITE_INDEX: 0,
+  FRAMES_AVAILABLE: 1,
+  LOCK: 2,
+}
+
+class WebsocketClient {
+  private readonly workerCtx: Worker
+  private targetSampleRate: number = 16000
+  private isContextStarted: boolean = false
+  private websocket?: WebSocket
+  private audioProcessor?: AudioProcessor
+  private controlSAB?: Int32Array
+  private dataSAB?: Float32Array
+
+  private readonly frameMillis = 30
+  private readonly outputAudioFrame: Int16Array = new Int16Array(this.frameMillis * this.targetSampleRate / 1000)
+
+  private debug: boolean = false
+
+  private defaultContextOptions?: ContextOptions
+
+  constructor(ctx: Worker) {
+    this.workerCtx = ctx
+  }
+
+  connect(apiUrl: string, authToken: string, targetSampleRate: number, debug: boolean): void {
+    this.debug = debug
+    if (this.debug) {
+      console.log('[WebSocketClient]', 'connecting to ', apiUrl)
+    }
+    this.targetSampleRate = targetSampleRate
+    this.isContextStarted = false
+    this.websocket = new WebSocket(apiUrl, authToken)
+    this.websocket.addEventListener('open', this.onWebsocketOpen)
+    this.websocket.addEventListener('message', this.onWebsocketMessage)
+    this.websocket.addEventListener('error', this.onWebsocketError)
+    this.websocket.addEventListener('close', this.onWebsocketClose)
+  }
+
+  initAudioProcessor(sourceSampleRate: number, vadOptions?: VadOptions): void {
+    this.audioProcessor = new AudioProcessor(sourceSampleRate, this.targetSampleRate, 5)
+
+    if (vadOptions) {
+      this.audioProcessor.vad = new EnergyTresholdVAD(vadOptions)
+
+      this.audioProcessor.onVadSignalHigh = () => {
+        const currentVadOptions = this.audioProcessor?.vad?.vadOptions
+        if (currentVadOptions) {
```

Comment:
5 levels of nesting is a bit hard to read, I suggest to return right away if `currentVadOptions` are not defined
```suggestion
        if (currentVadOptions === undefined) return
```

Reply:
Thanks for pointing this out. Did some cleanup there and it's much better now.